### ISSUE TITLE
disable REPL evaluation while terminal server is evaluating code

### ISSUE THREAD

**Issue by ghost** [JuliaEditorSupport/julia-vscode#105]
@davidanthoff  could you please try the following code

```
function run_code(code)
    repl = Base.active_repl
    julia_mod = repl.interface.modes[1]
    on_enter = julia_mod.on_enter
    on_done = julia_mod.on_done
    julia_mod.on_enter = (s) -> false
    julia_mod.on_done = (args...) -> nothing
    eval(parse(code))
    julia_mod.on_enter = on_enter
    julia_mod.on_done = on_done
    s = Base.LineEdit.init_state(repl.t, repl.interface)
    Base.LineEdit.refresh_multi_line(s)
end
```

What the code does is simple. First, before the terminal server evaluates code, it disables the "on_enter" and on _done methods of the julia prompt, which are activated when user press one or multiple '\n'. After evaluation, it changes the on_enter and on_done methods back.
You will see that the result is shown after the current user's text, and a new prompt. To go back editing just press back_space.

I haven't tested in the TerminalServer but it works in the REPL
For example, run 
`@asyn run_code("sleep(10); println(\"OK\")")`
and try to input something. After OK is printed, you can press back_space to get back to where you were.

Is this behavior what you're looking for?

**Comment by davidanthoff** [comment]
I think it is getting closer, but not yet there :)

Essentially what I want is this. Say the screen looks like this:
````
Some old output
julia> function foo(x)
   print
````
So here the user is in the middle of entering a new function at the REPL. Now say in this moment the background task receives code that it wants to run. The first thing it should do is delete line 2-3 from the screen completely, and position the cursor at (2,1). Then the code is run. Lets say that user code prints something with ``println``. At this point the screen should look like this:
````
Some old output
This was printed via println from the code
````
Once the user code is done running, the old REPL prompt should appear again, exactly like it was left of, so like this:
````
Some old output
This was printed via println from the code
julia> function foo(x)
   print
````
So the user can now continue editing the function in the REPL where he/she left things before the code on the background task came in.

I think the answer to this is to be found in [LineEdit.jl](https://github.com/JuliaLang/julia/blob/master/base/LineEdit.jl), but I don't really understand that code...

**Comment by novatena2015** [comment]
So it means, the background task will take over the out_stream of the terminal during its execution. During this time, the terminal in_stream is temporally disabled, the user doesn't see the code he's typing (maybe some kind of "BUSY, prompt" indicating that the background task is running). After its done, the background task releases the terminal in and out_stream and make the in_stream writable again.

**Comment by novatena2015** [comment]
It would be easy if we can access the current internal state of the active repl. But it seems it is impossible to do so :(

**Comment by davidanthoff** [comment]
Ah, that is *exactly* what we need. I wonder whether the approach in https://github.com/staticfloat/MakePkgUpdatePrecompileInTheBackground.jl could work in the meantime, until this is resolved in base... But they mention that one needs a special build of julia for that as well, so maybe not.

**Comment by ihnorton** [comment]
I have to say, between this issue and the -users thread, I'm still bewildered that you folks want to do this. I don't think I've ever seen this kind of behavior in an ide-integrated REPL. Maybe I'm missing something, but it seems both strange and unavoidably brittle.

**Comment by davidanthoff** [comment]
Well, if someone has a better idea, it would certainly be welcome. Here is what we are trying to achieve:

1. have a REPL in VS Code that is just the normal julia REPL. This we have working.
2. be able to send code from the editor for execution in that REPL process.

We do 2) by pasting the code into the REPL currently. But that means that the result of the expression that gets evaluated in the REPL will be shown in the REPL. Instead, I would like the results of the evaluation of code from the editor to be shown in the editor, ideally inline like in Juno, or maybe in a separate output tab or something like that. For that, we can't just paste the code into the REPL, we need to have a background task in the REPL process that listens for that code on a named pipe, and then executes it.

But if the code that from the editor that gets evaluated has any ``print`` statement in it, things will look messed up, because that output will now appear after the normal julia prompt. So instead, I was think we should hide the julia prompt while the background task is executing the code from the editor, and once that is done, show the REPL again.

If anyone has a better idea how we could solve that general workflow, I'd be all ears.

**Comment by ihnorton** [comment]
I see.

> But if the code that from the editor that gets evaluated has any print statement in it, things will look messed up, because that output will now appear after the normal julia prompt.

It should be possible to use `redirect_stdout` (and `_stderr`) to capture the output while the background task evaluates the code, and then hook up the original streams afterward.

**Comment by davidanthoff** [comment]
But any ``println`` of the background task actually should appear in the REPL (like it does e.g. in Juno right now), so I don't think just redirecting would work, right?

**Comment by ihnorton** [comment]
> But any println of the background task actually should appear in the REPL (like it does e.g. in Juno right now), so I don't think just redirecting would work, right?

Julia tasks can only do one thing at a time, so if you execute some long-running code on the "background task", the REPL should mostly just hang (unless you have a separate REPL and kernel, though this isn't quite supported by the existing code). I guess there might be a situation where code you executed in the "background task" yields, and another task started "from the REPL" decides to print, and intermixes. Is that the issue? I'm not sure if Juno handles that either, actually.

The `redirect_*` functions work by reassigning the global i/o handles, but it may be possible to do in a scoped way instead.

**Comment by davidanthoff** [comment]
I think the main issue is that if the background task prints, it will be appended to the julia prompt that is visible right now. Say the background task executes this line from the editor ``println("TEST")``, then what would show up in terminal is
````
julia> TEST
````
But really ``TEST`` should appear *before* the julia prompt, i.e. after the background task is done running, I want the terminal to look like this:
````
TEST
julia>
````
And if the user then enters some characters they should appear after the ``julia>`` part.

I think Juno gets around all of this pretty easily because they don't actually use the julia REPL but instead wrote their own thing from scratch with a completely different UI stack.

**Comment by ihnorton** [comment]
What I mean is that you can capture any output produced by the "background task" and display it in the IDE. It doesn't need to be shown in the REPL at all.

But as far as prepended output, I hacked up something that seems to mostly work for a minimal "prepend before current prompt" requirement (not nearly as fancy as what Elliott did for background precompile). I'll post it in the other thread, partly in hopes of baiting someone into improving it.

**Comment by davidanthoff** [comment]
> What I mean is that you can capture any output produced by the "background task" and display it in the IDE. It doesn't need to be shown in the REPL at all.

This won't work for VS Code. Unlike Atom, one can't just add arbitrary UI. Essentially there are a few well-defined extension points, one can use those, but add nothing else. Obviously is less flexible, but the upside is that VS Code is actually fast because MS controls all the UI... So we need a solution that works just in a normal terminal, because a julia terminal is one of the extension points VS Code has.

> But as far as prepended output, I hacked up something that seems to mostly work for a minimal "prepend before current prompt" requirement (not nearly as fancy as what Elliott did for background precompile). I'll post it in the other thread, partly in hopes of baiting someone into improving it.

Cool, thanks so much!

**Comment by ihnorton** [comment]
> This won't work for VS Code. Unlike Atom, one can't just add arbitrary UI.

Ok. I guess this is my confusion, because based on the comment below, I thought you could and wanted to do:

> Instead, I would like the results of the evaluation of code from the editor to be shown in the editor, ideally inline like in Juno, or maybe in a separate output tab or something like that.

**Comment by davidanthoff** [comment]
Well, I hope to do that eventually. The VS Code people say that displaying results of inline evaluations is high on their priority list, so I think there is hope that they will add another extension point that will enable that. But this would really only work for the result of an expression that gets evaluated, not for any printing that this code might do via ``println``, for that I do think it has to go to the REPL terminal.
